screengrab: init at 2016-02-24

romildo opened the pull request:
###### Things done
- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

romildo committed:
screengrab: init at 2016-02-24

romildo commented:
I am packaging [screengrab](https://github.com/QtDesktop/screengrab) (crossplatform tool for fast making screenshots, an optional application of the LXQt desktop).

Unfortunately the binary is not working:

```
$ result/bin/screengrab 
result/bin/screengrab: error while loading shared libraries: libuploader.so.1.0.0: cannot open shared object file: No such file or directory
```

I need some help in fixing this issue before complete the PR.

Mic92 commented on pkgs/desktops/lxqt/optional/screengrab/build-date-time.patch:
```diff
@@ -0,0 +1,22 @@
+-    versionInfo = tr("built on ");
+-    versionInfo.append(__DATE__);
+-    versionInfo.append(" ");
+-    versionInfo.append(__TIME__);
```
There modifications in nixos which replaces __DATE__ and __TIME__ by deterministic values. So this is not required.

Mic92 commented:
you need the following to get it working:

```patch
diff --git a/pkgs/desktops/lxqt/optional/screengrab/default.nix b/pkgs/desktops/lxqt/optional/screengrab/default.nix
index 370c741..2cf322b 100644
--- a/pkgs/desktops/lxqt/optional/screengrab/default.nix
+++ b/pkgs/desktops/lxqt/optional/screengrab/default.nix
@@ -35,6 +35,13 @@ stdenv.mkDerivation rec {
 
   cmakeFlags = [ "-DSG_USE_SYSTEM_QXT=ON" ];
 
+  postFixup = ''
+    target=$out/bin/screengrab
+    patchelf \
+      --set-rpath "$(patchelf --print-rpath "$target"):$out/lib/screengrab" \
+      "$target"
+  '';
+
   meta = with stdenv.lib; {
     description = "Crossplatform tool for fast making screenshots";
     homepage = https://github.com/lxde/screengrab;
```

romildo commented:
@Mic92 thanks for the fix.

For some days I have been reading about [`patchelf`](https://nixos.org/patchelf.html), as I knew the error message is related to the run-time path set in binaries. But I was not sure how to fix it.

In particular the article [http://sandervanderburg.blogspot.com.br/2015/10/deploying-prebuilt-binary-software-with.html](http://sandervanderburg.blogspot.com.br/2015/10/deploying-prebuilt-binary-software-with.html) has been helpful.

Now that I better understand it, I found a simpler fix by putting `-DCMAKE_INSTALL_LIBDIR=lib` into `cmakeFlags`.

Mic92 commented:
They have applied this pull request now: https://github.com/QtDesktop/screengrab/pull/29
Could you use `fetchPatch` to download the patch from: `https://github.com/QtDesktop/screengrab/commit/3dbacb9d6f52825689846c798a6c4c95e3815bf6.patch` instead. This makes future maintenance easier.

romildo commented:
I have updated the revision to `3dbacb9d6f52825689846c798a6c4c95e3815bf6` in `src`, so there is no need to explicitly patch. The two other patches from Debian are not that important. I have removed them too. The possibility of uploading is not shown to the user anymore.

romildo commented:
@Mic92 Thank you for helping.

Mic92 commented:
no problem.
